lib/spack/spack/fetch_strategy.py

```diff
@@ -730,6 +734,113 @@ def __str__(self):
         return "[go] %s" % self.url
 
 
+class InvalidGitRef(ValueError):
+    """Raised internally when a single git version can't be deduced."""
+
+
+class GitRef(object):
+    _known_types = ['commit', 'tag', 'branch']               # type: ClassVar[List[str]]
+    _ref_type = None                                         # type: str
+    _ref = None                                              # type: str
+
+    def __init__(self, ref_type, ref):
+        # type: (str, str) -> None
+        assert ref_type in self._known_types, (ref_type, self._known_types)
+        self._ref_type = ref_type
+        self._ref = ref
+
+    def eq_type(self, fn):
+        # type: (Any) -> bool
+        name = fn.__name__                                  # type: ignore[attr-defined]
+        assert name in self._known_types, (name, self._known_types)
+        return self._ref_type == name
+
+    def __repr__(self):
+        # type: () -> str
+        return '{0}(ref_type={1!r}, ref={2!r})'.format(type(self).__name__,
+                                                       self._ref_type, self._ref)
+
+    @classmethod
+    def commit(cls, ref):
+        # type: (str) -> GitRef
+        return cls('commit', ref)
+
+    @classmethod
+    def tag(cls, ref):
+        # type: (str) -> GitRef
+        return cls('tag', ref)
+
+    @classmethod
+    def branch(cls, ref):
+        # type: (str) -> GitRef
+        return cls('branch', ref)
+
+    # TODO: add test!
+    @classmethod
+    def from_directive(cls, commit, tag, branch, version_name):
+        # type: (Optional[str], Optional[str], Optional[str], Optional[str]) -> GitRef
+        num_specified = sum((x is not None) for x in [commit, tag, branch])
+        # If no explicit git ref arguments are provided, assume that the version string
+        # is to be used as the branch.
+        if (num_specified == 0) and version_name is not None:
+            return cls.branch(version_name)
+        # Otherwise, error if we have ambiguous ref arguments, or no ref arguments
+        # at all.
+        if num_specified != 1:
+            raise InvalidGitRef(dedent("""\
+            Exactly one of commit, tag, or branch must be specified for a git url.
+            Given:
+            commit={0}
+            tag={1}
+            branch={2}
+            """).format(commit, tag, branch))
+        if commit is not None:
+            return cls.commit(commit)
+        if tag is not None:
+            return cls.tag(tag)
+        assert branch is not None
+        return cls.branch(branch)
+
+    def refspec(self):
+        # type: () -> str
+        """Return a string which unambiguously identifies this ref's type."""
+        if self.eq_type(self.commit):
+            # TODO: validate that the commit is a sha256 checksum?
+            # Commits are always specified via hash, and the hash unambiguously
+            # specifies it.
+            return self._ref
+        if self.eq_type(self.tag):
```

Comment:
I believe your suggestion could actually make the code more difficult to read as well as significantly more prone to silent errors that could eat up a ridiculous amount of someone's time in the future, and I would like to avoid you, me, or anyone else having to spend that time in the future, hence the lengthy discussion below.

Someone else reading this code is already going to have to stop and understand the inherent complexity of the implementation, since we're fundamentally changing how we cache git repos. Here's what they would also have to think about if we applied your suggestion here:
```python
if self._ref_type == 'tag':
    ...
```

1. what is `self._ref_type`? It's a private variable, so could be anything.
1. what is `'tag'`? Is that a valid value for `self._ref_type` here? Did I spell the string correctly?
1. what does it mean for `self._ref_type` to be equal to `'tag'` here? Where is that described? What if that meaning changes later?

In contrast, the current code allows me to observe at a glance, without spooky action at a distance:
```python
if self.eq_type(self.tag):
    ...
```

1. `self.eq_type()` is checking what type of reference the current `self` is. This is clear from the name and from the fact that it is used everywhere else in this class.
1. `self.tag` will never, ever be misspelled, because mypy will error before runtime for a nonexistent attribute, and python will error at runtime. Both errors make it extremely clear what mistake was made.
1. Being equal to `self.tag` means we are the variant of `GitRef` which corresponds to a git tag. I know this because it is directly drawn from the name of the sole classmethod which is used in this file and in testing to *construct* tag refs.

The benefit we're seeing from this suggestion that would make up for the significantly increased risk of comparing private variables against string values is not clear to me. Just by *using* the method `self.eq_type()`, we get all of the benefits above, and we now can avoid having to flip back to the class's constructor every time we want to know what a variable is.

This is akin to the rationale for isolating this conditional logic into the `GitRef` class in the first place. We're not adding type checking, we are using an abstraction which lets us perform much more semantically-meaningful operations, such as calculating the argument to provide to git to update the specific refspec, without having to wonder whether I've correctly checked that `self` is supposed to be a `tag`.

Separately from anything else, I want to understand how we would respond to the case of a misspelled string, and how a user would be expected to diagnose that, and then I'd like to revisit the expected benefit of this suggestion to use `self._ref_type == 'tag'` instead of `self.eq_type(self.tag)`.